library/core/src/char/decode.rs

```diff
@@ -120,9 +120,21 @@ impl<I: Iterator<Item = u16>> Iterator for DecodeUtf16<I> {
     #[inline]
     fn size_hint(&self) -> (usize, Option<usize>) {
         let (low, high) = self.iter.size_hint();
-        // we could be entirely valid surrogates (2 elements per
-        // char), or entirely non-surrogates (1 element per char)
-        (low / 2, high)
+
+        // `self.buf` will never contain the first part of a surrogate,
```

Comment:
Why? It doesn't seem to me like that's the case.

For example the following would fail the test below.

```rust
check(&[0xD800, 0xD800, 0xDC00]);
```

```console
thread 'char::test_decode_utf16_size_hint' panicked at 'lower = 2, upper = Some(2)', library/core/tests/char.rs:320:13
```

Reply:
It was a wrong assumption from the original PR that I haven't checked 😅 

I pushed a fix that checks the contents of the buf.